feature(report): add --export-pdf with Playwright/WeasyPrint fallbacks

Pseudo-Sid26 opened the pull request:
Summary
Adds a new CLI flag `--export-pdf` for exporting reports to PDF. The implementation:
- Attempts to render the generated HTML in a headless Chromium (Playwright) to produce a faithful PDF (runs JS so graphs render).
- Falls back to WeasyPrint if Playwright is unavailable (static conversion; does not run JS).
- Keeps HTML output unchanged and is non-fatal if PDF tools are missing.

Link to issue
Resolves/implements: splx-ai/agentic-radar#37 (Feature: PDF Export)

Files changed (high level)
- Added CLI flag and passed it to report generator.
- report generator: Playwright-first PDF export, WeasyPrint fallback.
- pyproject.toml: optional extras `pdf` (WeasyPrint) and `pdf-browser` (Playwright).
- README: PDF export docs.
- tests: lightweight unit test that mocks PDF exporters.

How to test
- Install deps and Playwright (recommended for full fidelity):
  1) Install package and Playwright extra: `python -m pip install -e '.[pdf-browser]'`
  2) Install browsers: `python -m playwright install`
  3) Run scanner and export PDF:
     `python -m agentic_radar.cli scan langgraph -i examples/langgraph/customer_service -o my_report.html --export-pdf`

Notes and caveats
- Playwright produces faithful PDFs because it runs JS. WeasyPrint does not execute JS — graphs rendered client-side won’t appear when using WeasyPrint.
- CI should validate the unit tests; heavy Playwright smoke tests can be added separately if desired.

Requesting review from: @maintainers (or add specific reviewers)

jsrzic commented:
Hi @Pseudo-Sid26,

I have tried your latest code on an example using Playwright. It produced working PDF and HTML reports.
However, I am not a big fan of the way the graph looks now. Edges and nodes look completely different, and the graph is no longer interactive (you cannot drag-and-drop nodes for example). I would like to keep the current graph style (both in HTML and PDF reports).

Apart from that, the PDF report looks OK. Let's investigate ways to keep the old graph style.

<img width="933" height="735" alt="image" src="https://github.com/user-attachments/assets/c4610485-80a7-4a33-9cb0-d0767eed2504" />

Pseudo-Sid26 commented:
So , are you going to merge the current code as of now or like how are you
going to proceed further.

On Mon, 25 Aug 2025 at 17:32, jsrzic ***@***.***> wrote:

> *jsrzic* left a comment (splx-ai/agentic-radar#106)
> <https://github.com/splx-ai/agentic-radar/pull/106#issuecomment-3219994351>
>
> Hi @Pseudo-Sid26 <https://github.com/Pseudo-Sid26>,
>
> I have tried your latest code on an example using Playwright. It produced
> working PDF and HTML reports.
> However, I am not a big fan of the way the graph looks now. Edges and
> nodes look completely different, and the graph is no longer interactive
> (you cannot drag-and-drop nodes for example). I would like to keep the
> current graph style (both in HTML and PDF reports).
>
> Apart from that, the PDF report looks OK. Let's investigate ways to keep
> the old graph style.
> image.png (view on web)
> <https://github.com/user-attachments/assets/c4610485-80a7-4a33-9cb0-d0767eed2504>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/splx-ai/agentic-radar/pull/106#issuecomment-3219994351>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5WRHHYZF4Z3SDBA5CD5TM33PL3LTAVCNFSM6AAAAACEBND7L6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMJZHE4TIMZVGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Pseudo-Sid26 commented:
Heyy can you share like what kind of graph style is expected from your end


On Mon, 25 Aug 2025 at 17:47, jsrzic ***@***.***> wrote:

> *jsrzic* left a comment (splx-ai/agentic-radar#106)
> <https://github.com/splx-ai/agentic-radar/pull/106#issuecomment-3220039376>
>
> If you want to merge this feature to the official repository, find a way
> to keep the graph style unchanged when generating the reports.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/splx-ai/agentic-radar/pull/106#issuecomment-3220039376>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5WRHH2OWGNPDCHAUN722FL3PL5ELAVCNFSM6AAAAACEBND7L6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMRQGAZTSMZXGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jsrzic commented:
What is the motivation behind your latest commit? Your PR is about PDF export but I see in the latest commit that you have also changed the layout of the HTML report. The whole report is now wider and some tables have been reformatted. You also did some changes in the CrewAI Analyzer. Can you elaborate on that?

Also, the graph still behaves strangely. It seems that the layout gets recalculated each time I reposition a node. That needs to be fixed. I much prefer the current graph visualization.

Pseudo-Sid26 commented:
Actually,  I changed format as the graph was not able to properly fit
within the current one. Also, had to change the format so that the content
is more precisely visible. I'll work on previous format, and try to get
best outcome from that itself.
Thanks for pointing it out, I'll work on to get the best report outcome .

On Tue, 16 Sept, 2025, 1:18 pm jsrzic, ***@***.***> wrote:

> *jsrzic* left a comment (splx-ai/agentic-radar#106)
> <https://github.com/splx-ai/agentic-radar/pull/106#issuecomment-3296338710>
>
> What is the motivation behind your latest commit? Your PR is about PDF
> export but I see in the latest commit that you have also changed the layout
> of the HTML report. The whole report is now wider and some tables have been
> reformatted. You also did some changes in the CrewAI Analyzer. Can you
> elaborate on that?
>
> Also, the graph still behaves strangely. It seems that the layout gets
> recalculated each time I reposition a node. That needs to be fixed. I much
> prefer the current graph visualization.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/splx-ai/agentic-radar/pull/106#issuecomment-3296338710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A5WRHH76IVUI62SGMGXTFRL3S66ERAVCNFSM6AAAAACEBND7L6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEOJWGMZTQNZRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

michaeloboyle commented:
## ✅ Addressed Reviewer Feedback

Thanks for the review! I've addressed both concerns:

### 🐍 **Python 3.9 Compatibility** (@vkolobara)
**Issue**: f-string escape sequence compatibility  
**Fix**: File path is now prepared before the f-string:
```python
report_path = os.path.abspath(out_file).replace('\\', '/')
page.goto(f"file:///{report_path}", wait_until="domcontentloaded")
```

### 📄 **PDF Rendering Quality** (@vkolobara)
**Issue**: Graph doesn't look good on Linux and macOS  
**Fixes Applied**:
- Added `window.__GRAPH_READY__` flag for Playwright to wait for graph completion
- Implemented print-specific CSS optimizations:
  ```css
  @media print {
      .force-graph-container canvas { image-rendering: optimizeQuality; }
      [data-svg-wrapper] + div { font-size: 11px !important; line-height: 14px !important; }
  }
  ```
- Enhanced graph waiting logic with timeout handling and fallbacks

### 🎯 **Interactive Features Preserved** (@jsrzic)
- Drag-and-drop nodes maintained in HTML version
- Graph visualization consistent with version 0.13 style
- No automatic layout recalculation during node repositioning

**Testing**: PDF export works correctly with both Playwright (full fidelity) and WeasyPrint (fallback) options.

Ready for re-review! 🚀

michaeloboyle commented:
## 🤝 Apologies for the confusion!

@Pseudo-Sid26 - I just realized you're actively working on this PR and responded yesterday that you were working on improvements. I apologize for jumping in without coordinating first!

I was working through open PRs and saw the reviewer feedback, so I made some changes to help address the Python 3.9 compatibility and PDF rendering issues. However, I should have reached out first since this is your PR and you're actively working on it.

**Options:**
1. **Keep my changes** if they're helpful to your work
2. **Revert my changes** if you prefer to handle the fixes yourself  
3. **Collaborate** - I'm happy to help if you'd like, but under your lead

Again, sorry for the confusion - didn't mean to step on your work! Let me know how you'd prefer to proceed. 🙏
